js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx

```diff
@@ -59,170 +63,151 @@ export const WorkspaceContext = React.createContext<WorkspaceState>(
-      ... on RepositoryLocation {
-        id
-        ...WorkspaceLocation
-      }
-      ...PythonErrorFragment
-    }
-  }
+  const {data} = codeLocationStatusQueryResult;
 
-  fragment WorkspaceDisplayMetadata on RepositoryMetadata {
-    key
-    value
-  }
+  const locations = useMemo(() => getLocations(data), [data]);
+  const prevLocations = useRef<typeof locations>({});
 
-  fragment WorkspaceLocation on RepositoryLocation {
-    id
-    isReloadSupported
-    serverId
-    name
-    dagsterLibraryVersions {
-      name
-      version
-    }
-    repositories {
-      id
-      ...WorkspaceRepository
-    }
-  }
+  const didInitiateFetchFromCache = useRef(false);
+  const [didLoadCachedData, setDidLoadCachedData] = useState(false);
 
-  fragment WorkspaceRepository on Repository {
-    id
-    name
-    pipelines {
-      id
-      name
-      isJob
-      isAssetJob
-      pipelineSnapshotId
-    }
-    schedules {
-      id
-      ...WorkspaceSchedule
-    }
-    sensors {
-      id
-      ...WorkspaceSensor
-    }
-    partitionSets {
-      id
-      mode
-      pipelineName
-    }
-    assetGroups {
-      id
-      groupName
-    }
-    allTopLevelResourceDetails {
-      id
-      name
-    }
-    ...RepositoryInfoFragment
-  }
+  const [locationsData, setLocationsData] = React.useState<
+    Record<string, WorkspaceLocationNodeFragment | PythonErrorFragment>
+  >({});
 
-  fragment WorkspaceSchedule on Schedule {
-    id
-    cronSchedule
-    executionTimezone
-    mode
-    name
-    pipelineName
-    scheduleState {
-      id
-      selectorId
-      status
+  useEffect(() => {
+    // Load data from the cache
+    if (didInitiateFetchFromCache.current) {
+      return;
     }
-  }
+    didInitiateFetchFromCache.current = true;
+    (async () => {
+      const data = await getCachedData<CodeLocationStatusQuery>({
+        key: `${localCacheIdPrefix}/${CODE_LOCATION_STATUS_QUERY_KEY}`,
+        version: CODE_LOCATION_STATUS_QUERY_VERSION,
+      });
+      const cachedLocations = getLocations(data);
+      const prevCachedLocations: typeof locations = {};
+
+      await Promise.all([
+        ...Object.values(cachedLocations).map(async (location) => {
+          const locationData = await getCachedData<LocationWorkspaceQuery>({
+            key: `${localCacheIdPrefix}/${locationWorkspaceKey(location.name)}`,
+            version: LOCATION_WORKSPACE_QUERY_VERSION,
+          });
+          const entry = locationData?.workspaceLocationEntryOrError;
+
+          if (!entry) {
+            return;
+          }
+          setLocationsData((locationsData) =>
+            Object.assign({}, locationsData, {
+              [location.name]: entry,
+            }),
+          );
+
+          if (entry.__typename === 'WorkspaceLocationEntry') {
+            prevCachedLocations[location.name] = location;
+          }
+        }),
+      ]);
+      prevLocations.current = prevCachedLocations;
+      setDidLoadCachedData(true);
+    })();
```

Comment:
I stared at this for a while - I think it deserves a short comment explaining that even though we are not awaiting the promise it will keep progressing and the eventual state set calls provide the required control flow

Reply:
The only reason we don't await the promise is because useEffect functions can't be async. Adding a comment though since it isn't obvious why this is necessary without looking at the code further below